### PR TITLE
Improve Enum.TryParse performance. Fixes #70

### LINKED ISSUE #70: Parsing string to enum with digit as first char takes a very long time
From https://github.com/dotnet/corefx/issues/594

When parsing strings to enums with TryParse, in some cases this takes way longer than in all the other cases. After a bit of research I found that all the strings with a digit as first char are slower. This is due to the attempt to parse the string to the underlying type (https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Enum.cs#L396) with a Convert.ChangeType, which can throw an exception, because further down the line it uses (for example) Int32.Parse instead of TryParse (https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Convert.cs#L1118).

The time it takes to create the Exception is significant. In my project I ended up writing a new Enum parser to circumvent this issue.
"Something": 240.49 ns per parse
"1Somethin": 37178.57 ns per parse
"123456789": 485.63 ns per parse

As users expect Convert.ToXxx() to throw an exception when it fails, changing this is a bad idea I think. Possible solutions I can think of:
- Extend the Convert class with Try-methods that don't throw exceptions, and use those while trying to parse an enum (much work)
- Instead of checking whether the first character is a digit (https://github.com/Microsoft/referencesource/blob/master/mscorlib/system/enum.cs#L434), change that line to a TryParse of the whole string (the easier way)
- One of the options @Alexx99 posted in the other issue (https://github.com/dotnet/corefx/issues/594#issuecomment-72893658)

### PR DESCRIPTION
**Pull request by Alexx999**
The idea here is that enum value is always integer, so it will surely fit into long if it's signed or into ulong if it's unsigned.  Furthermore, string-to-enum code uses ulong to store both signed and unsigned results anyway, so there is no real need to make any difference between signed/unsigned other than handling negative numbers or cast to intermediate results.

Performance figures:
Since I was unable to run MeasureIt on CoreClr, not was I able to find System.dll to get Stopwatch this results was taken using DateTime.Now and are fairly approximate, but difference is so big that it is hard to make mistake here.

Measurements taken on i5-4670K CPU at 4.2GHz, Win 8.1,  release build
Old:
```
Good run: 218ms
Bad run: 13657ms
Big numbers run: 10955ms
Small numbers run: 295ms
All letters run: 125ms
```
New:
```
Good run: 215ms
Bad run: 187ms
Big numbers run: 252ms
Small numbers run: 218ms
All letters run: 123ms
```

In worst case improvement are over 70 times, plus there is noticeable speedup of parsing numbers as such (something near 25% since we now do less operations)

Code I used to test this can be viewed as gist: https://gist.github.com/Alexx999/84640c2c31e4f5806cd1

Issues/Open questions:

I found one edge case with undefined behavior - more precisely, when you do supply value that overflows enum's backing data type. This case is not covered by tests in CoreFx, it is present in my test as "Case 13".
For example, imagine following code:
``` C#
enum SmallEnum : byte
{}
bool result = Enum.TryParse(ulong.MaxValue.ToString(), out smallEnum);
```
In this case you'll get false with old code (and new code will parse it fine).

I believe this is inconsistent with how enums work in general - you can do this:
``` C#
ulong bigValue = ulong.MaxValue;
SmallEnum smallEnum = (SmallEnum)bigValue;
```
and it will not fail. Even more than that - this behavior actually *is* covered by tests, as can be seen here:
https://github.com/dotnet/corefx/blob/master/src/System.Runtime/tests/System/Enum.cs#L290

So is it okay to have such behavior changes or it must be changed to perfectly mimic old behavior?

Fixes #70

### CODE REVIEW

**Comment by AlexGhiondea** [comment]
Thanks for the contribution.

Since this code is shared with desktop we need to make sure we don't break compatibility with it. 

Cases where the behavior is different before/after the change are  considered breaking changes and need to be mitigated. 

There are a couple of ways we mitigate breaking changes:
- the code can be if-def'ed to only impact CoreCLR
- we can code the fix such that it behaves one way on older platforms (ie. current version of Desktop) but get the new behavior for newer platforms. (this is handled by/with the AppContext class).

For this specific case, you can probably special case the situation where the enum backing type is smaller than the value that you need to assign to it and in that case have the old behavior.

**Comment by Alexx999** [comment]
OK, I got it - we play safe here :)
I've got one idea how to make it behave exactly like the old one without hurting performance too much, I'll make it work that way

**Comment by Alexx999** [comment]
Implemented that change.
New performance figures:
```
Good run: 214ms
Bad run: 187ms
Big numbers run: 443ms
Small numbers run: 319ms
All letters run: 121ms
```
It is now a bit slower with large numbers (still 25 times faster than original) and has roughly equal performance with small numbers.

**Comment by AlexGhiondea** [comment]
Thanks for the contribuition. I am going to run some validation and if everything is ok I'll merge this in.

**Comment by nguerrera** [comment]
Note that bool is a valid enum underlying type. How does this change interact with that? 

e.g. Given:
```
.class public sealed PeculiarEnum extends [mscorlib]System.Enum
{
  .field public specialname rtspecialname bool value__
  .field public static literal valuetype PeculiarEnum A = bool(false)
  .field public static literal valuetype PeculiarEnum B = bool(true)
}
```
Does the behavior of  `Enum.Parse("0", typeof(PeculiarEnum))` change?

At a glance, I suspect it would have thrown FormatException before, but return PeculiarEnum.A now.

**Comment by Alexx999** [comment]
@nguerrera yeah, you're right.
I've updated test with this particular case - I think we should update Enum tests in CoreFx too.

I think this particular case can be worked around with additional check.
```
Good run: 213ms
Bad run: 182ms
Big numbers run: 482ms
Small numbers run: 356ms
All letters run: 122ms
```

**Comment by Alexx999** [comment]
Updated test to check if char-based enum works correctly.
It turns out that in code
```C#
Enum.Parse("0", typeof(CharEnum))
```
"0" was treated like a char and not like number, so the result must be
```C#
Enum.ToObject(typeof(CharEnum), '0')
```
and it was equal to
```C#
Enum.ToObject(typeof(CharEnum), 0)
```

Now new code basically mimics behavior of Convert.ChangeType approach, but without throwing.
As a bonus - it's now faster compared to previous solution with extra checks.

```
Good run: 214ms
Bad run: 203ms
Big numbers run: 247ms
Small numbers run: 229ms
All letters run: 124ms
```

File: src/mscorlib/src/System/Enum.cs [hunk 1]
```diff
@@ -386,30 +386,27 @@ private static bool TryParseEnum(Type enumType, String value, bool ignoreCase, r
             // values will have the first character as as number or a sign.
             ulong result = 0;
 
-            if (Char.IsDigit(value[0]) || value[0] == '-' || value[0] == '+')
+            if (Char.IsDigit(value[0]) || value[0] == '+' || value[0] == '-')
```
Comment (AlexGhiondea) [inline src/mscorlib/src/System/Enum.cs]: Any particular reason for switching these statements?

File: src/mscorlib/src/System/Enum.cs [hunk 2]
```diff
@@ -386,30 +386,27 @@ private static bool TryParseEnum(Type enumType, String value, bool ignoreCase, r
...
-                    if (parseResult.canThrow)
-                        throw;
-                    else
+                    try
                     {
-                        parseResult.SetFailure(ex);
-                        return false;
+                        parseResult.parsedEnum = ToObject(enumType, result);
+                        return true;
+
```
Comment (AlexGhiondea) [inline src/mscorlib/src/System/Enum.cs]: Extra line.

**Comment by AlexGhiondea** [comment]
Overall the change looks good. I left some comments in the PR. 
Could you squash the commits into a single one?

If nothing comes out of the test run I am doing I'll merge this in!

File: src/mscorlib/src/System/Enum.cs (hunk 1 above)
Comment (Alexx999) [inline src/mscorlib/src/System/Enum.cs]: No, that wasn't intended.

**Comment by Alexx999** [comment]
@AlexGhiondea OK, I cleaned up remaining things and squashed it to one commit.

**Comment by AlexGhiondea** [comment]
Thanks for all the work you did on this change. I have kicked off a different test run and should have results on Monday. The test run I did didn't show up any issues.

However, I talked with @jkotas about this and, to be on the safe side, we should update the code so that in the default case of the switch statement we fall back to the slow behavior we had before. This would cover the bool/float/double case.

Can you also check what happens when we have an enum backed by a float/double? I know you cannot create these in C#, but via IL or C++ they could still be generated and we need to make sure we can correctly handle them.

Again, thank you for doing the due diligence around this fix.

**Comment by Alexx999** [comment]
@AlexGhiondea Indeed, @jkotas is right - there is possibility of having enum backed by float/double/IntPtr/UIntPtr.
This can be seen here:
https://github.com/dotnet/coreclr/blob/9df15b9fdfb55ce2076a5e7eede8ca71481e5e3b/src/mscorlib/src/System/Enum.cs#L749
But on the same time ToObject() method doesn't support any of those types (as opposed to bool and char - they're actually supported) so parsing it will fail anyway, see
https://github.com/dotnet/coreclr/blob/9df15b9fdfb55ce2076a5e7eede8ca71481e5e3b/src/mscorlib/src/System/Enum.cs#L550

**Comment by Alexx999** [comment]
Okay, added one more fixup that will ensure correct behavior in all cases - it will throw ArgumentException with same message as before.
Performance seems to be unaffected, at least difference is less than measurement error.
```
Good run: 212ms
Bad run: 204ms
Big numbers run: 245ms
Small numbers run: 237ms
All letters run: 129ms
```

**Comment by AlexGhiondea** [comment]
@Alexx999 I would like to thank you for the work you put into this PR.

We have updated our contribution guidelines to better call out the bar for accepting pull requests. You can find it here: https://github.com/dotnet/coreclr/wiki/Contribution-guidelines

We are aware of the work you put into this PR and we truly appreciate it! But, unfortunately, since it is not a mainline scenario and has unknown risk, this PR does not meet the bar.

We are sorry for not being able to accept it and hope you will continue making contributions!

Thank you again for your contribution!

**Comment by Phrohdoh** [comment]
It is quite sad to see a PR like this being discarded after all of the work put forward here.

**Comment by Alexx999** [comment]
that's a bit of surprise - it just seemed like all possible cases are covered at last.

but probably you are right - this case is not really common.

**Comment by AlexGhiondea** [comment]
@Phrohdoh I know.

One thing I would like to do is, with @Alexx999's permission, capture the tests in the CoreFX repo (I can make the change or review a PR). 
That way, if we ever touch that area in the future we have a better understanding of the tricky parts.

**Comment by weshaggard** [comment]
@Phrohdoh I agree it is unfortunate as well and I also appreciate the effort that @Alexx999 put into this and while we were having the conversation on the PR we were also having a conversation internally to try and figure out the appropriate bar, hopefully that is a little clearer now. In this case while it would be nice to take the reward isn't worth the potential risk. 

Thanks again @alex999 and I look forward to your future contributions.

**Pull request closed without merging**